src/dmd/backend/inliner.d

```diff
@@ -0,0 +1,482 @@
+            {   // Never inline expand this function
+
+                // But do expand templates
+                Symbol* s = e1.EV.E1.EV.Vsym;
+                if (tyfunc(s.ty()))
+                {
+                    // This function might be an inline template function that was
+                    // never parsed. If so, parse it now.
+                    if (s.Sfunc.Fbody)
+                    {
+                        //n2_instantiate_memfunc(s);
+                    }
+                }
+            }
+            else
+                e1.EV.E1 = inline_do_walk(e1.EV.E1);
+            e1.EV.E2 = inline_do_walk(e1.EV.E2);
+        }
+        else
+        {
+            e.EV.E1 = inline_do_walk(e.EV.E1);
+            if (op == OPucall)
+            {
+                e = inline_ifcan(e);
+            }
+        }
+    }
+    else /* leaf */
+    {
+        // If deferred allocation of variable, allocate it now.
+        // The deferred allocations are done by cpp_initctor().
+        if (0 && CPP &&
+            (op == OPvar || op == OPrelconst))
+        {
+            Symbol* s = e.EV.Vsym;
+            if (s.Sclass == SCauto &&
+                s.Ssymnum == SYMIDX.max)
+            {   //dbg_printf("Deferred allocation of %p\n",s);
+                symbol_add(s);
+
+                if (tybasic(s.Stype.Tty) == TYstruct &&
+                    s.Stype.Ttag.Sstruct.Sdtor &&
+                    !(s.Sflags & SFLnodtor))
+                {
+                    elem* eptr = el_ptr(s);
+                    elem* edtor ; //= cpp_destructor(s.Stype,eptr,null,DTORmostderived);
+                    assert(edtor);
+                    edtor = inline_do_walk(edtor);
+                    //cpp_stidtors.push(edtor);
+                }
+            }
+            if (tyfunc(s.ty()))
+            {
+                // This function might be an inline template function that was
+                // never parsed. If so, parse it now.
+                if (s.Sfunc.Fbody)
+                {
+                    //n2_instantiate_memfunc(s);
+                }
+            }
+        }
+    }
+    return e;
+}
+
+/**********************************
+ * Inline-expand a function call if it can be.
+ * Params:
+ *      e = OPcall or OPucall elem
+ * Returns:
+ *      replacement tree.
+ */
+
+private elem* inline_ifcan(elem *e)
+{
+    //elem_debug(e);
+    assert(e && (e.Eoper == OPcall || e.Eoper == OPucall));
+
+    if (e.EV.E1.Eoper != OPvar)
+        return e;
+
+    // This is an explicit function call (not through a pointer)
+    Symbol* sfunc = e.EV.E1.EV.Vsym;
+    if (log) printf("inline_ifcan: %s, class = %d\n", prettyident(sfunc),sfunc.Sclass);
+
+    // sfunc may not be a function due to user's clever casting
+    if (!tyfunc(sfunc.Stype.Tty))
+        return e;
+
+    /* If forward referencing an inline function, we'll have to
+     * write out the function when it eventually is defined
+     */
+    if (sfunc.Sfunc.Fstartblock == null)
```

Comment:
Segfault here:
```
 ... compilable/test14317.d         -O -profile -inline  -fPIC ()
==============================
Test 'compilable/test14317.d' failed. The logged output:
/home/ibuclaw/src/dmd/dmd/generated/linux/release/64/dmd -conf= -m64 -Icompilable -O -profile -inline  -fPIC  -odtest_results/compilable -oftest_results/compilable/test14317_0.o  -c compilable/test14317.d 
Segmentation fault (core dumped)

==============================
Test 'compilable/test14317.d' failed: Expected rc == 0, but exited with rc == 139

 ... compilable/test19320.d         -O -cov  -fPIC ()
==============================
Test 'compilable/test19320.d' failed. The logged output:
/home/ibuclaw/src/dmd/dmd/generated/linux/release/64/dmd -conf= -m64 -Icompilable -O -cov  -fPIC  -odtest_results/compilable -oftest_results/compilable/test19320_0.o  -c compilable/test19320.d 
Segmentation fault (core dumped)

==============================
Test 'compilable/test19320.d' failed: Expected rc == 0, but exited with rc == 139

 ... compilable/test15898.d         -O  -fPIC ()
==============================
Test 'compilable/test15898.d' failed. The logged output:
/home/ibuclaw/src/dmd/dmd/generated/linux/release/64/dmd -conf= -m64 -Icompilable -O  -fPIC  -odtest_results/compilable -oftest_results/compilable/test15898_0.o  -c compilable/test15898.d 
Segmentation fault (core dumped)

==============================
Test 'compilable/test15898.d' failed: Expected rc == 0, but exited with rc == 139

 ... compilable/test17146.d         -O -inline  -fPIC ()
==============================
Test 'compilable/test17146.d' failed. The logged output:
/home/ibuclaw/src/dmd/dmd/generated/linux/release/64/dmd -conf= -m64 -Icompilable -O -inline  -fPIC  -odtest_results/compilable -oftest_results/compilable/test17146_0.o  -c compilable/test17146.d 
Segmentation fault (core dumped)

==============================
Test 'compilable/test17146.d' failed: Expected rc == 0, but exited with rc == 139

 ... compilable/b15206.d            -O  -fPIC ()
==============================
Test 'compilable/b15206.d' failed. The logged output:
/home/ibuclaw/src/dmd/dmd/generated/linux/release/64/dmd -conf= -m64 -Icompilable -O  -fPIC  -odtest_results/compilable -oftest_results/compilable/b15206_0.o  -c compilable/b15206.d 
Segmentation fault (core dumped)

==============================
Test 'compilable/b15206.d' failed: Expected rc == 0, but exited with rc == 139

 ... compilable/test20990.d         -O  -fPIC ()
==============================
Test 'compilable/test20990.d' failed. The logged output:
/home/ibuclaw/src/dmd/dmd/generated/linux/release/64/dmd -conf= -m64 -Icompilable -O  -fPIC  -odtest_results/compilable -oftest_results/compilable/test20990_0.o  -c compilable/test20990.d 
Segmentation fault (core dumped)

==============================
Test 'compilable/test20990.d' failed: Expected rc == 0, but exited with rc == 139

 ... compilable/ccompile.d          -O  -fPIC ()
==============================
Test 'compilable/ccompile.d' failed. The logged output:
/home/ibuclaw/src/dmd/dmd/generated/linux/release/64/dmd -conf= -m64 -Icompilable -O  -fPIC  -odtest_results/compilable -oftest_results/compilable/ccompile_0.o  -c compilable/ccompile.d 
Segmentation fault (core dumped)

==============================
Test 'compilable/ccompile.d' failed: Expected rc == 0, but exited with rc == 139

 ... compilable/testVRP.d            -fPIC (-O -inline)
==============================
Test 'compilable/testVRP.d' failed. The logged output:
/home/ibuclaw/src/dmd/dmd/generated/linux/release/64/dmd -conf= -m64 -Icompilable  -fPIC  -odtest_results/compilable -oftest_results/compilable/testVRP_0.o  -c compilable/testVRP.d 

/home/ibuclaw/src/dmd/dmd/generated/linux/release/64/dmd -conf= -m64 -Icompilable  -fPIC -O -odtest_results/compilable -oftest_results/compilable/testVRP_1.o  -c compilable/testVRP.d 
Segmentation fault (core dumped)

==============================
Test 'compilable/testVRP.d' failed: Expected rc == 0, but exited with rc == 139
```

Reply:
Backtrace:
```
Program received signal SIGSEGV, Segmentation fault.
0x0000555555af0f3e in dmd.backend.inliner.inline_ifcan(dmd.backend.el.elem*) (e=0x5555560170b0) at src/dmd/backend/inliner.d:255
255	    if (sfunc.Sfunc.Fstartblock == null)
#0  0x0000555555af0f3e in dmd.backend.inliner.inline_ifcan(dmd.backend.el.elem*) (e=0x5555560170b0) at src/dmd/backend/inliner.d:255
#1  0x0000555555af0dc5 in dmd.backend.inliner.inline_do_walk(dmd.backend.el.elem*) (e=0x5555560170b0) at src/dmd/backend/inliner.d:150
#2  0x0000555555af0dab in dmd.backend.inliner.inline_do_walk(dmd.backend.el.elem*) (e=0x555556020bd0) at src/dmd/backend/inliner.d:148
#3  0x0000555555af0d94 in dmd.backend.inliner.inline_do_walk(dmd.backend.el.elem*) (e=0x5555560060d0) at src/dmd/backend/inliner.d:147
#4  0x0000555555af0d94 in dmd.backend.inliner.inline_do_walk(dmd.backend.el.elem*) (e=0x555556032510) at src/dmd/backend/inliner.d:147
#5  0x0000555555af0d94 in dmd.backend.inliner.inline_do_walk(dmd.backend.el.elem*) (e=0x5555560171d0) at src/dmd/backend/inliner.d:147
#6  0x0000555555af0d03 in dmd.backend.inliner.inline_do(dmd.backend.cc.Symbol*) (sfunc=0x555556050950) at src/dmd/backend/inliner.d:125
#7  0x0000555555ae0dc4 in optfunc() () at src/dmd/backend/go.d:342
#8  0x0000555555af03a5 in writefunc2(Symbol*) (sfunc=0x555556050950) at src/dmd/backend/out.d:1386
#9  0x0000555555aefd9f in writefunc(Symbol*) (sfunc=0x555556050950) at src/dmd/backend/out.d:1138
#10 0x0000555555a93309 in FuncDeclaration_toObjFile(FuncDeclaration*, bool) (fd=0x7ffff7b9e500, multiobj=false) at src/dmd/glue.d:1186
#11 0x0000555555aab5ae in toObjFile::ToObjFile::visit(FuncDeclaration*) (this=0x7fffffffc9f8, fd=0x7ffff7b9e500) at src/dmd/toobj.d:315
#12 0x00005555559b874e in FuncDeclaration::accept(Visitor*) (this=0x7ffff7b9e500, v=0x7fffffffc9f8) at src/dmd/func.d:2920
#13 0x0000555555aab515 in toObjFile(Dsymbol*, bool) (ds=0x7ffff7b9e500, multiobj=false) at src/dmd/toobj.d:1022
#14 0x0000555555a9153c in genObjFile(Module*, bool) (m=0x7ffff7b73550, multiobj=false) at src/dmd/glue.d:523
#15 0x0000555555a8fd8b in dmd.glue.generateCodeAndWrite(dmd.dmodule.Module[], const(char)*[], const(char)[], const(char)[], bool, bool, bool, bool, bool) (verbose=false, multiobj=false, oneobj=true, obj=true, lib=false, objdir=..., libname=..., libmodules=..., modules=...) at src/dmd/glue.d:121
#16 0x00005555558a2bd3 in dmd.mars.tryMain(ulong, const(char)**, ref dmd.globals.Param) (params=..., argv=0x7fffffffd858, argc=10) at src/dmd/mars.d:578
#17 0x00005555558a4152 in D main (_param_0=...) at src/dmd/mars.d:962
```

Reply:
Looks like these are compiler-generated run-time library helpers.
```
(gdb) call prettyident(sfunc)
$1 = 0x555555e94b24 "_d_arraybounds_indexp"
```
i.e: https://github.com/dlang/dmd/blob/15d8805b40edff7f8fdaa7d103316f4a85f1cf8c/src/dmd/e2ir.d#L6692

`el_var` does not set `Sfunc`.